### PR TITLE
Add aishell recipe

### PR DESCRIPTION
**Pull request by pkufool**
Add aishell recipe to icefall. It contains two models now, one is phone based Tdnn-Lstm model, the other is char based (with chinese chars as modelling units) comformer-ctc model. 

~~The decoding results seems not as good as expected now, I am still tuning the models.~~

TODOs:
- [x] update results and documents.
- [x] add pre-trained models

### CODE REVIEW

**Comment by fanlu** [comment]
I have an error with the newest master branch of k2.
```
2021-09-19 09:40:39,362 INFO [decode.py:419] Decoding started
2021-09-19 09:40:39,362 INFO [decode.py:420] {'exp_dir': PosixPath('conformer_ctc/exp'), 'lang_dir': PosixPath('data/lang_char'), 'lm_dir': PosixPath('data/lm'), 'subsampling_factor': 4, 'feature_dim': 80, 'nhead': 4, 'attention_dim': 512, 'num_en$
oder_layers': 12, 'num_decoder_layers': 6, 'vgg_frontend': False, 'use_feat_batchnorm': True, 'search_beam': 20, 'output_beam': 7, 'min_active_states': 30, 'max_active_states': 10000, 'use_double_scores': True, 'epoch': 40, 'avg': 18, 'method': 'a$
tention-decoder', 'num_paths': 100, 'lattice_score_scale': 0.5, 'export': False, 'feature_dir': PosixPath('data/fbank'), 'max_duration': 20, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.$
, 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2}
2021-09-19 09:40:39,633 INFO [lexicon.py:113] Loading pre-compiled data/lang_char/Linv.pt
2021-09-19 09:40:39,834 INFO [decode.py:430] device: cuda:0
2021-09-19 09:40:50,195 INFO [decode.py:470] averaging ['conformer_ctc/exp/epoch-23.pt', 'conformer_ctc/exp/epoch-24.pt', 'conformer_ctc/exp/epoch-25.pt', 'conformer_ctc/exp/epoch-26.pt', 'conformer_ctc/exp/epoch-27.pt', 'conformer_ctc/exp/epoch-2$
.pt', 'conformer_ctc/exp/epoch-29.pt', 'conformer_ctc/exp/epoch-30.pt', 'conformer_ctc/exp/epoch-31.pt', 'conformer_ctc/exp/epoch-32.pt', 'conformer_ctc/exp/epoch-33.pt', 'conformer_ctc/exp/epoch-34.pt', 'conformer_ctc/exp/epoch-35.pt', 'conformer$
ctc/exp/epoch-36.pt', 'conformer_ctc/exp/epoch-37.pt', 'conformer_ctc/exp/epoch-38.pt', 'conformer_ctc/exp/epoch-39.pt', 'conformer_ctc/exp/epoch-40.pt']
2021-09-19 09:41:17,977 INFO [decode.py:483] Number of model parameters: 115125888
[F] /asr_storage/fanlu/k2/k2/csrc/array.h:501:void k2::Array1<T>::Init(k2::ContextPtr, int32_t, k2::Dtype) [with T = int; k2::ContextPtr = std::shared_ptr<k2::Context>; int32_t = int] Check failed: size >= 0 (-119123572 vs. 0) Array size MUST be g$
eater than or equal to 0, given :-119123572


[ Stack-Trace: ]
/asr_storage/fanlu/k2/build_release/lib/libk2_log.so(k2::internal::GetStackTrace()+0x3c) [0x2af6a59029fc]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::internal::Logger::~Logger()+0x2a) [0x2af6a40b037a]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Array1<int>::Init(std::shared_ptr<k2::Context>, int, k2::Dtype)+0x26d) [0x2af6a40c56dd]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x25a19d) [0x2af6a425619d]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x272d92) [0x2af6a426ed92]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Index(k2::RaggedShape&, int, k2::Array1<int> const&, k2::Array1<int>*)+0x1b2) [0x2af6a4270eb2]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x147977) [0x2af6a329e977]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x1197b0) [0x2af6a32707b0]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10be3d) [0x2af6a3262e3d]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10c08e) [0x2af6a326308e]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x24663) [0x2af6a317b663]
```

**Comment by danpovey** [comment]
Hm.  Try doing:
export  CUDA_LAUNCH_BLOCKING=1
and see if you can get a Python stack trace.

**Comment by danpovey** [comment]
.. and if you can't get a Python stack trace, then:
 (a) I'd like to figure out what is wrong with our assert-failure mechanism
that we can't get this,
 (b) to debug the immediate issue, you may have to insert debug print
statements to figure out where we are-- some kind of binary search into the
code.

**Comment by danpovey** [comment]
... Another way to debug this might be to run it in gdb, as in
gdb --args python3 decode.py ...
  (gdb) r
and see if you can get any info from the C++ stack.

**Comment by csukuangfj** [comment]
> I have an error with the newest master branch of k2.
> 
> ```
> 
> 2021-09-19 09:40:39,362 INFO [decode.py:419] Decoding started
> 
> 2021-09-19 09:40:39,362 INFO [decode.py:420] {'exp_dir': PosixPath('conformer_ctc/exp'), 'lang_dir': PosixPath('data/lang_char'), 'lm_dir': PosixPath('data/lm'), 'subsampling_factor': 4, 'feature_dim': 80, 'nhead': 4, 'attention_dim': 512, 'num_en$
> 
> oder_layers': 12, 'num_decoder_layers': 6, 'vgg_frontend': False, 'use_feat_batchnorm': True, 'search_beam': 20, 'output_beam': 7, 'min_active_states': 30, 'max_active_states': 10000, 'use_double_scores': True, 'epoch': 40, 'avg': 18, 'method': 'a$
> 
> tention-decoder', 'num_paths': 100, 'lattice_score_scale': 0.5, 'export': False, 'feature_dir': PosixPath('data/fbank'), 'max_duration': 20, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.$
> 
> , 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2}
> 
> 2021-09-19 09:40:39,633 INFO [lexicon.py:113] Loading pre-compiled data/lang_char/Linv.pt
> 
> 2021-09-19 09:40:39,834 INFO [decode.py:430] device: cuda:0
> 
> 2021-09-19 09:40:50,195 INFO [decode.py:470] averaging ['conformer_ctc/exp/epoch-23.pt', 'conformer_ctc/exp/epoch-24.pt', 'conformer_ctc/exp/epoch-25.pt', 'conformer_ctc/exp/epoch-26.pt', 'conformer_ctc/exp/epoch-27.pt', 'conformer_ctc/exp/epoch-2$
> 
> .pt', 'conformer_ctc/exp/epoch-29.pt', 'conformer_ctc/exp/epoch-30.pt', 'conformer_ctc/exp/epoch-31.pt', 'conformer_ctc/exp/epoch-32.pt', 'conformer_ctc/exp/epoch-33.pt', 'conformer_ctc/exp/epoch-34.pt', 'conformer_ctc/exp/epoch-35.pt', 'conformer$
> 
> ctc/exp/epoch-36.pt', 'conformer_ctc/exp/epoch-37.pt', 'conformer_ctc/exp/epoch-38.pt', 'conformer_ctc/exp/epoch-39.pt', 'conformer_ctc/exp/epoch-40.pt']
> 
> 2021-09-19 09:41:17,977 INFO [decode.py:483] Number of model parameters: 115125888
> 
> [F] /asr_storage/fanlu/k2/k2/csrc/array.h:501:void k2::Array1<T>::Init(k2::ContextPtr, int32_t, k2::Dtype) [with T = int; k2::ContextPtr = std::shared_ptr<k2::Context>; int32_t = int] Check failed: size >= 0 (-119123572 vs. 0) Array size MUST be g$
> 
> eater than or equal to 0, given :-119123572
> 
> 
> 
> 
> 
> [ Stack-Trace: ]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2_log.so(k2::internal::GetStackTrace()+0x3c) [0x2af6a59029fc]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::internal::Logger::~Logger()+0x2a) [0x2af6a40b037a]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Array1<int>::Init(std::shared_ptr<k2::Context>, int, k2::Dtype)+0x26d) [0x2af6a40c56dd]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x25a19d) [0x2af6a425619d]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x272d92) [0x2af6a426ed92]
> 
> /asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Index(k2::RaggedShape&, int, k2::Array1<int> const&, k2::Array1<int>*)+0x1b2) [0x2af6a4270eb2]
> 
> /asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x147977) [0x2af6a329e977]
> 
> /asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x1197b0) [0x2af6a32707b0]
> 
> /asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10be3d) [0x2af6a3262e3d]
> 
> /asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10c08e) [0x2af6a326308e]
> 
> /asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x24663) [0x2af6a317b663]
> 
> ```

There are instructions at the end of log, telling you how to use gdb to debug your Python code.

**Comment by csukuangfj** [comment]
Are there Python stack traces at the end of the log that are not posted?

**Comment by fanlu** [comment]
@csukuangfj the end of the log is
```
python conformer_ctc/decode.py --lattice-score-scale 0.5                                --epoch 40                                --avg 18                                --method attention-decoder
                      --max-duration 20                                --num-paths 100
2021-09-19 09:40:39,362 INFO [decode.py:419] Decoding started
2021-09-19 09:40:39,362 INFO [decode.py:420] {'exp_dir': PosixPath('conformer_ctc/exp'), 'lang_dir': PosixPath('data/lang_char'), 'lm_dir': PosixPath('data/lm'), 'subsampling_factor': 4, 'feature_dim': 80, 'nhead': 4, 'attention_dim': 512, 'num_enc
oder_layers': 12, 'num_decoder_layers': 6, 'vgg_frontend': False, 'use_feat_batchnorm': True, 'search_beam': 20, 'output_beam': 7, 'min_active_states': 30, 'max_active_states': 10000, 'use_double_scores': True, 'epoch': 40, 'avg': 18, 'method': 'at
tention-decoder', 'num_paths': 100, 'lattice_score_scale': 0.5, 'export': False, 'feature_dir': PosixPath('data/fbank'), 'max_duration': 20, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0
, 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2}
2021-09-19 09:40:39,633 INFO [lexicon.py:113] Loading pre-compiled data/lang_char/Linv.pt
2021-09-19 09:40:39,834 INFO [decode.py:430] device: cuda:0
2021-09-19 09:40:50,195 INFO [decode.py:470] averaging ['conformer_ctc/exp/epoch-23.pt', 'conformer_ctc/exp/epoch-24.pt', 'conformer_ctc/exp/epoch-25.pt', 'conformer_ctc/exp/epoch-26.pt', 'conformer_ctc/exp/epoch-27.pt', 'conformer_ctc/exp/epoch-28
.pt', 'conformer_ctc/exp/epoch-29.pt', 'conformer_ctc/exp/epoch-30.pt', 'conformer_ctc/exp/epoch-31.pt', 'conformer_ctc/exp/epoch-32.pt', 'conformer_ctc/exp/epoch-33.pt', 'conformer_ctc/exp/epoch-34.pt', 'conformer_ctc/exp/epoch-35.pt', 'conformer_
ctc/exp/epoch-36.pt', 'conformer_ctc/exp/epoch-37.pt', 'conformer_ctc/exp/epoch-38.pt', 'conformer_ctc/exp/epoch-39.pt', 'conformer_ctc/exp/epoch-40.pt']
2021-09-19 09:41:17,977 INFO [decode.py:483] Number of model parameters: 115125888
[F] /asr_storage/fanlu/k2/k2/csrc/array.h:501:void k2::Array1<T>::Init(k2::ContextPtr, int32_t, k2::Dtype) [with T = int; k2::ContextPtr = std::shared_ptr<k2::Context>; int32_t = int] Check failed: size >= 0 (-119123572 vs. 0) Array size MUST be gr
eater than or equal to 0, given :-119123572


[ Stack-Trace: ]
/asr_storage/fanlu/k2/build_release/lib/libk2_log.so(k2::internal::GetStackTrace()+0x3c) [0x2af6a59029fc]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::internal::Logger::~Logger()+0x2a) [0x2af6a40b037a]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Array1<int>::Init(std::shared_ptr<k2::Context>, int, k2::Dtype)+0x26d) [0x2af6a40c56dd]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x25a19d) [0x2af6a425619d]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(+0x272d92) [0x2af6a426ed92]
/asr_storage/fanlu/k2/build_release/lib/libk2context.so(k2::Index(k2::RaggedShape&, int, k2::Array1<int> const&, k2::Array1<int>*)+0x1b2) [0x2af6a4270eb2]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x147977) [0x2af6a329e977]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x1197b0) [0x2af6a32707b0]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10be3d) [0x2af6a3262e3d]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x10c08e) [0x2af6a326308e]
/asr_storage/fanlu/k2/build_release/lib/_k2.cpython-38-x86_64-linux-gnu.so(+0x24663) [0x2af6a317b663]
python(PyCFunction_Call+0x58) [0x558070ade348]
python(_PyObject_MakeTpCall+0x23c) [0x558070acddbc]
python(+0x19a74c) [0x558070b2474c]
python(+0x10343c) [0x558070a8d43c]
python(_PyEval_EvalCodeWithName+0x300) [0x558070b23270]
python(_PyFunction_Vectorcall+0x1e3) [0x558070b240a3]
python(+0x10343c) [0x558070a8d43c]
python(_PyEval_EvalCodeWithName+0x8b1) [0x558070b23821]
python(_PyFunction_Vectorcall+0x1e3) [0x558070b240a3]
python(+0x10343c) [0x558070a8d43c]
python(_PyEval_EvalCodeWithName+0x300) [0x558070b23270]
python(_PyFunction_Vectorcall+0x1e3) [0x558070b240a3]
python(+0x10343c) [0x558070a8d43c]
python(_PyFunction_Vectorcall+0x10b) [0x558070b23fcb]
python(PyObject_Call+0x414) [0x558070ace754]
python(_PyEval_EvalFrameDefault+0x1fe6) [0x558070b570b6]
python(_PyEval_EvalCodeWithName+0x8b1) [0x558070b23821]
python(_PyFunction_Vectorcall+0x1e3) [0x558070b240a3]
python(+0x1035db) [0x558070a8d5db]
python(_PyEval_EvalCodeWithName+0x300) [0x558070b23270]
python(PyEval_EvalCode+0x23) [0x558070bb8543]
python(+0x22e5e4) [0x558070bb85e4]
python(+0x254854) [0x558070bde854]
python(+0x115390) [0x558070a9f390]
python(PyRun_SimpleFileExFlags+0x384) [0x558070aa20d2]
python(+0x118bf0) [0x558070aa2bf0]
python(Py_BytesMain+0x39) [0x558070be1a09]
/usr/lib64/libc.so.6(__libc_start_main+0xf5) [0x2af5752c8495]
python(+0x1e6fe5) [0x558070b70fe5]

Traceback (most recent call last):
  File "conformer_ctc/decode.py", line 515, in <module>
    main()
  File "/asr_storage/fanlu/miniconda3/envs/k2_cu111/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "conformer_ctc/decode.py", line 494, in main
    results_dict = decode_dataset(
  File "conformer_ctc/decode.py", line 330, in decode_dataset
    hyps_dict = decode_one_batch(
  File "conformer_ctc/decode.py", line 266, in decode_one_batch
    best_path_dict = rescore_with_attention_decoder(
  File "/mnt/cfs1_alias2/asr/users/fanlu/task/icefall/icefall/decode.py", line 826, in rescore_with_attention_decoder
    token_seq, _ = token_seq.index(
RuntimeError:
    Some bad things happened. Please read the above error messages and stack
    trace. If you are using Python, the following command may be helpful:

      gdb --args python /path/to/your/code.py

    (You can use `gdb` to debug the code. Please consider compiling
    a debug version of k2.).

    If you are unable to fix it, please open an issue at:

      https://github.com/k2-fsa/k2/issues/new
```

**Comment by csukuangfj** [comment]
>   File "/mnt/cfs1_alias2/asr/users/fanlu/task/icefall/icefall/decode.py", line 826, in rescore_with_attention_decoder
    token_seq, _ = token_seq.index(
RuntimeError:

Could you set a **pdb** breakpoint at line 826 and print the shape information of `token_seq` and `new2old`?

Some useful instructions are:

```python3
import pdb
pdb.set_trace()
print(token_seq.shape.tot_sizes())
print(new2old.shape)
# token_seq, _ = token_seq.index(...
```

**Comment by fanlu** [comment]
> > File "/mnt/cfs1_alias2/asr/users/fanlu/task/icefall/icefall/decode.py", line 826, in rescore_with_attention_decoder
> > token_seq, _ = token_seq.index(
> > RuntimeError:
> 
> Could you set a **pdb** breakpoint at line 826 and print the shape information of `token_seq` and `new2old`?
> 
> Some useful instructions are:
> 
> ```python
> import pdb
> pdb.set_trace()
> print(token_seq.shape.tot_sizes())
> print(new2old.shape)
> # token_seq, _ = token_seq.index(...
> ```

image

**Comment by csukuangfj** [comment]
Does this `token_seq` cause the above runtime error?
Could you also show
```
print(new2old.max(), new2old.min())
```

**Comment by fanlu** [comment]
> Does this `token_seq` cause the above runtime error?
> Could you also show
> 
> ```
> print(new2old.max(), new2old.min())
> ```

image
yes,`token_seq.index` cause the above runtime error

**Comment by csukuangfj** [comment]
We resolved it offline.

It is a bug that has been fixed in https://github.com/k2-fsa/icefall/pull/50

File: egs/aishell/ASR/prepare.sh [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  # The directory structure is
+  # aishell/
+  # |-- data_aishell
+  # |   |-- transcript
+  # |   `-- wav
+  # `-- resource_aishell
+  #     |-- lexicon.txt
+  #     `-- speaker.info
+
+  if [ ! -d $dl_dir/aishell/wav ]; then
```
Comment (pkufool) [inline egs/aishell/ASR/prepare.sh:72]: $dl_dir/aishell/data_aishell/wav

**Comment by pkufool** [comment]
Latest result:
image

Tensorboard: https://tensorboard.dev/experiment/WE1DocDqRRCOSAgmGyClhg/#scalars

image

**Comment by csukuangfj** [comment]
Could you also update https://github.com/k2-fsa/icefall/blob/master/README.md ?

**Pull request merged**